activerecord/lib/active_record/associations/builder/has_one.rb

```diff
@@ -17,5 +17,12 @@ def self.valid_dependent_options
     def self.add_destroy_callbacks(model, reflection)
       super unless reflection.options[:through]
     end
+
+    def self.define_validations(model, reflection)
+      super
+      if reflection.options[:required]
+        model.validates_presence_of reflection.name, message: :required
+      end
+    end
```

Comment:
I think it'll be odd to switch to `optional:` on belongs_to but keep `required:` for has_one.